### DISCUSSION TITLE
Typing Components using the 'as' prop

### DISCUSSION THREAD

**Discussion by jeffweinberg** [ariakit/ariakit#835] [Questions] [answered]
Hello - 

I am looking for some assistance in typing a component that uses the `as` prop. I have come across @diegohaz example at [https://codesandbox.io/s/typescript-as-prop-dicj8](https://codesandbox.io/s/typescript-as-prop-dicj8) but I am still having trouble extracting a pattern for typing a Reakit based component that will understand what props are expected from the component defined in the `as` prop.

Here is a specific scenario - I am defining a "Dropdown" button that allows the user to pass a custom component using `as` or it will default to a Button. 

```
 import { Button,  MenuButton as ReakitMenuButton } from 'reakit';

// I want to know how to replace any here with proper typing
export const DropdownMenuButton = ({ as, ...htmlProps }: any) => {
  const { menu } = useDropdownMenuContext();
  return (
    <ReakitMenuButton {...menu}>
      {(menuProps) => {
        let buttonProps = { ...htmlProps, ...menuProps };
        return as ? as(buttonProps) : <Button {...buttonProps} />;
      }}
    </ReakitMenuButton>
  );
};

```

Now - if the custom component requires a `name` prop, there is currently no way for Typescript to know this:

```
const CustomTriggerButton = (props: { name: string }) => (
  <div {...props}>
    hello from component
  </div>
);

// This works without a TS error (missing the name prop):

<DropdownMenuButton as={CustomTriggerButton}>
    Custom button
</DropdownMenuButton>

```

Here is a [Code Sandbox](https://codesandbox.io/s/as-prop-demo-8pr8p?file=/src/App.tsx) that shows this in action. Thank you for your help!

**Comment by jeffweinberg** [comment]
@diegohaz - are there any docs you could point me to? Thank you for your help

**Comment by diegohaz** [comment, marked as answer]
Hello @jeffweinberg!

You can replace `ButtonProps` and `LinkProps` from the CodeSandbox example by the types from Reakit (like `MenuButtonProps`) or extend them.

**Reply by sseagull** [reply]
Ah, I think the crux of this problem is that @jeffweinberg doesn't know the component that will be used as the value to `as` ahead of time. The challenge is how to type `DropdownMenuButton` in a way that says:

If I have an `as` prop component, determine what props THAT component takes, and allow them as 'my' props as well

```tsx
const CustomButton = ({ uniqueProp }) => <button />

const DropdownMenuButton = ({ as, ...props }) => as ? as(props) : <button {...props} />

// when DropdownMenuButton is used like this, we want DropdownMenuButton to 'know' that its allowed to take `uniqueProp`
// because thats a prop on CustomButton
<DropdownMenuButton as={CustomButton} uniqueProp={'value'}  />
```

**Reply by diegohaz** [reply]
That's exactly what the CodeSandbox example does. `Link` has a custom `isPrimary` prop. You're able to use `<Button as={Link} isPrimary />`. So you just need to replace `ButtonProps` and `LinkProps` with whatever custom props you want, which could be Reakit specific types, but not necessarily. For example: https://codesandbox.io/s/typescript-as-prop-with-reakit-5qnvr?file=/src/App.tsx:1357-1911

**Reply by diegohaz** [reply]
Another note: you should pass `{...htmlProps}` to the `ReakitMenuButton` component instead of merging it into the other props inside the render prop, otherwise you may have props overridden.

```jsx
    <ReakitMenuButton {...menu} {...htmlProps}>
      {(menuProps) => {
        return as ? <As {...menuProps}>{htmlProps.children}</As> : <Button {...menuProps}>{htmlProps.children}</Button>;
      }}
    </ReakitMenuButton>
```

Also, `as(props)` is different from `<As {...props} />`. If you're using the `as` prop, you should use the latter version.

**Reply by sseagull** [reply]
Ooo thanks for the follow up. The sandbox should be enough to go off!

> Also, as(props) is different from <As {...props} />. If you're using the as prop, you should use the latter version.

Good call, I didn't know I could do that

**Reply by diegohaz** [reply]
I explain some of the differences between `as(props)` (which is actually render props) and `<As {...props} />` (the actual `as` prop, or component injection) here: https://www.youtube.com/watch?v=_kyv3tnQKfg&ab_channel=ReactFinland

**Reply by jeffweinberg** [reply]
Thank you, this is very helpful!
